### DISCUSSION TITLE
Is there any way to use NextAuth in class based component?

### DISCUSSION THREAD

**Discussion by MateuszWawrzynski** [nextauthjs/next-auth#840] [Help] [answered]
**Your question**
Is there any way to use NextAuth in class based component? Maybe through props passing or something?

**What are you trying to do**
My app components are made using ES6 classes. As we know - we can't use hooks there. I have to implement this library to make auth system in this app, but that makes me to rebuild most of my app to function components.

**Feedback**
*-*

* [ ] Found the documentation helpful
* [x] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [ ] Found the example project helpful
* [x] Did not find the example project helpful

**Comment by balazsorban44** [comment, marked as answer]
You could create a HOC (Higher Order Component) that sends them as props to your class component:

```jsx
import {useSession} from "next-auth/client"

const withSession = ClassComponent => props => {
  const session = useSession()
  
  if(ClassComponent.prototype.render) { // if the component has a render property, we are good
    return <ClassComponent session={session} {...props}/>
  }

  // if the passed component is a Function Component, there is no need for this wrapper
  throw new Error([
   "You passed a function component, `withSession` is not needed.",
   "You can `useSession` directly in your component."
  ].join("\n"))
}

// Usage

class ClassComponent extends React.Component {
  render() {
    const {session} = this.props
    return null
  }
}

const FuncComponent = ({session}) => {
  const session = useSession() // use this
  return null
}

const ClassComponentWithSession = withSession(ClassComponent)

const FuncComponentWithSession = withSession(FuncComponent) // will throw error, as this is not needed
```

**Reply by krisztianzagonyi-mintestio** [reply]
@balazsorban44 Can you explain a little bit more? With exports and other things please? It doesn't work for me. I would like to create a component, which can be exported to every sites, and check te session. :)

**Reply by balazsorban44** [reply]
Which part is unclear?

If you have a class component, simply wrap it with `withSession`. If you use function components, you won't need `withSession` at all, just use [`useSession`](https://next-auth.js.org/getting-started/client#usesession) directly.

Does this help?:
```js
// class-component.jsx
class ClassComponent extends React.Component {
  ...
  render() {
    console.log(this.props.session) // Equals to what `useSession` would return in function components
    return "Foo"
  }
}
export default withSession(ClassComponent)

// index.js
import ClassComponent from "./class-component"

export default function SomeComponent() {
  return (<ClassComponent/>)
}
```

To be honest, if you don't work with a legacy codebase, I would avoid class components altogether! It complicates things so much, and since we've got hooks there is little to no good reason to use them for new code.

**Reply by krisztianzagonyi-mintestio** [reply]
Thanks for your reply, and I am using legacy codebase. :(
I did exactly what you write here, I got the "Foo" on the site, but in the terminal I get 1 console log with session undefined, and in the browser console, I get first undefined session, than 2 console.log with session. It's like the browser loads before gets the session!?

**Reply by balazsorban44** [reply]
It's probably because of the hydration step. Nextjs renders a static version of the page. That gets sent to the user. When that is loaded, you'll get the undefined session in the terminal. Then when the interactive code is hydrated, the loading of the session starts. first it is undefined, (and loading is set to true) then it gets downloaded through the /api/auth/session endpoint, and this.props.session becomes the actual value.

**Reply by krisztianzagonyi-mintestio** [reply]
Exactly. Do you have any idea how can I prevent this?

**Reply by balazsorban44** [reply]
You can't, it's just Next.js. Don't see a reason why you would struggle preventing it anyway.

**Reply by krisztianzagonyi-mintestio** [reply]
Thanks for your help. 👍🏽 I have more to learn. :D

**Reply by balazsorban44** [reply]
I mean I might be wrong myself, but I don't think it is something to worry about. If you start seeing performance issues, then you can start debugging the cause, and if you end up here again, I'll gladly try to help :slightly_smiling_face:
